lib/ccflay.rb

```diff
@@ -0,0 +1,86 @@
+require "flay"
+require "concurrent"
+require "digest"
+
+##
+# A thread-safe and stable hash subclass of Flay.
+
+class CCFlay < Flay
+  def initialize(option = nil)
+    super
+
+    @hashes = Concurrent::Hash.new do |hash, key|
+      hash[key] = Concurrent::Array.new
+    end
+
+    self.identical = Concurrent::Hash.new
+    self.masses = Concurrent::Hash.new
+  end
+
+  ##
+  # Calculate the structural hash for this sexp. Cached, so don't
+  # modify the sexp afterwards and expect it to be correct.
+
+  if ENV["PURE_HASH"]
```

Comment:
No, but it'll allow us to A/B test them w/ a very minor ENV change. The pure version seems to be much nicer to memory footprint at the least. Benchmarks faster but that doesn't seem to affect wall clock on real projects (not the bottleneck on bigger projects).